lib/train/transports/azure_helpers/file_credentials.rb

```diff
@@ -0,0 +1,39 @@
+require 'inifile'
+require 'train/transports/azure_helpers/file_parser'
+require 'train/transports/azure_helpers/subscription_number_file_parser'
+require 'train/transports/azure_helpers/subscription_id_file_parser'
+
+module Train
+  module Transports
+    module AzureHelpers
+      class FileCredentials
+        DEFAULT_FILE = ::File.join(Dir.home, '.azure', 'credentials')
+
+        def self.parse(subscription_id: nil, credentials_file: DEFAULT_FILE, **_)
+          credentials     = IniFile.load(::File.expand_path(credentials_file))
+          subscription_id = parser(subscription_id, ENV['AZURE_SUBSCRIPTION_NUMBER'], credentials).subscription_id
+          creds(subscription_id, credentials)
+        end
+
+        def self.parser(subscription_id, subscription_number, credentials)
+          if subscription_id
+            SubscriptionIdFileParser.new(subscription_id, credentials)
+          elsif !subscription_number.nil?
+            SubscriptionNumberFileParser.new(subscription_number.to_i, credentials)
+          else
+            FileParser.new(credentials)
+          end
+        end
+
+        def self.creds(subscription_id, credentials)
+          {
+            subscription_id: subscription_id,
+            tenant_id:       credentials[subscription_id]['tenant_id'],
+            client_id:       credentials[subscription_id]['client_id'],
+            client_secret:   credentials[subscription_id]['client_secret']
```

Comment:
Either these should be `fetch` operations or there should be a `rescue NoMethodError` statement for trying to call `[]` on `nil`. Unless you're *really sure* all those values exist before getting here. =^)

Reply:
`credentials[subscription_id]` should be safe. If we use either the 0 index method or the index based lookup the subscription id is read from the credentials file. If find the credentials with the subscription_id then we have a guard to ensure that section exists in the file. 

There may be an edge case if the file has no entries. I can test that out.